Fixes for migrated Categories add/edit actions

sarjon opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | First PR of Categories migration was merged because it became huge, but a lot of minor fixes were needed.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10194
| How to test?  | See fixed ticket: page `/admin-dev/index.php/sell/catalog/categories/` listing migration is fixed. HOWEVER there are 2 known issues (see below) that will be fixed in another PR (waiting for other PRs to be merged). There was some early checks by marionf so QA should be easy for this page 😄 

### Known issues/todo list:
- [x] Make rich text editor (TinyMCE) translatable
- [x] Add `CleanHtml` constraint to description (waiting for https://github.com/PrestaShop/PrestaShop/pull/12163 to be merged)
- [x] Add Name to LinkRewrite copier (waiting for https://github.com/PrestaShop/PrestaShop/pull/12717 to be merged)
- [x] Language behavior for inputs is being fixed at https://github.com/PrestaShop/PrestaShop/pull/12214

### Testing
:warning: Assets needs to be built before testing.

Extra use cases:
1. When MultiShop feature is used, adding root category should be enabled.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12782)
<!-- Reviewable:end -->

sarjon committed:
add meta keywords input with tags

sarjon committed:
add cancel button to category form

sarjon committed:
validate legacy category fields when adding new category

sarjon committed:
fix category parent choice tree

sarjon committed:
preselect parent when creating new category

sarjon committed:
add form validations to Category form

sarjon committed:
fix grid alignment

sarjon commented on src/Core/Form/IdentifiableObject/DataProvider/CategoryFormDataProvider.php:
```diff
@@ -88,6 +95,7 @@ public function getDefaultData()
         $defaultGroups = $this->queryBus->handle(new GetDefaultGroups());
```
@matks we have 2 ways of getting default groups:

1. By using Query like above.
2. By using `\PrestaShop\PrestaShop\Adapter\Group\Provider\DefaultGroupsProvider` service

What would be the correct way? :thinking: we have to get rid of one of these 2.

matks replied:
A bit hard to answer because this is the borderline of CQRS. I think since this is kind of configuration we can go for solution 2, using DefaultGroupsProvider. But please add a `@todo` statement on it saying something like "Not sure whether or not to use a Query here". This way later we'll know where to look at, and we might be able to validate this choice using the knowledge and experience acquired in the meantime.

sarjon commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -97,28 +112,69 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'required' => false,
                 'options' => [
-                    'max_length' => 70,
+                    'max_length' => 255,
```
i don't like hardcoding this constraint into form, any suggestion for the right place to extract it?

matks replied:
I think i't's fine 😄 because this is a Form. It's display configuration, and it's in the right place.
You can put it in a constant at the top of the class but I'm not sure it will bring a lot of value.

Extracting a number is useful if you reuse this number in your app. But 255 is an arbitratry setting here. I dont think it will be targeted elsewhere.

sarjon replied:
i've copied it from `Category` object model `$definition`, so it's used for validation, we could extract that.

sarjon committed:
fix getting shop association for position update

sarjon committed:
redirect to parent category after editing

sarjon commented:
@matks currently images submitted via form are uploaded in command handler. However, I feel that Domain layer should not be responsible for uploading images, thus I suggest moving image uploading into `FormHandler`, just after Category is saved using `Command` and removing command dependency on `UploadedFile` Symfony's object, wdyt?

sarjon committed:
implement text with length counter for meta descirption

sarjon committed:
fix validation for TextWithLengthCounterType input

matks commented:
> I feel that Domain layer should not be responsible for uploading images

File management and file upload are borderline 😬 I'm never sure whether or not they belong to Domain, not matter the system I'm working on ...

@eternoendless an opinion on this ?

sarjon committed:
move command properties to AddCategoryCommand from abstract command

sarjon committed:
remove abstract category command

sarjon committed:
optimize adding/editing category

sarjon committed:
optimize adding/editing root category

sarjon commented:
hey @marionf , do you think you can help me on testing Categories Add/Edit actions? This page is pretty hard. :/

If so, there are few known issues/todos that are mentioned in PR description, so you can ignore that.

sarjon committed:
run cs fixer

marionf commented:
Hey @sarjon, yes sure, I take a look :)

marionf commented:
@sarjon I have the exception below on categories listing, so can't try add/edit actions

![capture d ecran_1134](https://user-images.githubusercontent.com/13449658/53946354-dce05300-40c3-11e9-89c6-19f459c571b7.png)

sarjon commented:
@marionf weird, it works well for me, could you try clearing cache from Back Office?

also, ill try to rebase my branch.

sarjon committed:
fix route name in category grid definition

sarjon commented:
@marionf after rebase i managed to reproduce that issue & fixed it. :+1:

marionf commented:
Thank you @sarjon 

Here are some returns:

1) On the add/edit page, tooltips are missing for the following fields: name, displayed, description, category cover image, category thumbnail, menu thumbnails, meta title, meta description, meta keywords, friendly url and group acces.

2) French is my default language (in International > Localization)
I select French in the add/edit category page
I miss to fill a mandatory field (like name)
Click on Save
I have well the error message but the language selected is now English instead of French.

![capture d ecran_1140](https://user-images.githubusercontent.com/13449658/53959404-d6fa6a00-40e3-11e9-83d8-6cd6fa86b2d0.png)

3) When I edit a category name & save, it's not saved

4) When I add a new category in the home category, I have 2 categories with position 1

![capture d ecran_1142](https://user-images.githubusercontent.com/13449658/53962677-f8f7ea80-40eb-11e9-8314-ad0664f8396a.png)

5) I am in multistore context and select the context "All shops"
When I click on add a new category, only the first shop is selected

![capture d ecran_1143](https://user-images.githubusercontent.com/13449658/53963278-41fc6e80-40ed-11e9-866e-c25a9c82a8db.png)

6) When I click to edit the "home" category I am redirected to the legacy page

![capture d ecran_1144](https://user-images.githubusercontent.com/13449658/53963415-8f78db80-40ed-11e9-9421-650ae27e2efe.png)

Hope it can help you, if you find some weird behavior, let me know

sarjon committed:
fix name saving for categories

sarjon committed:
fix categories position generation

sarjon committed:
fixed default values for shop association

sarjon committed:
fix url to edit Root category

sarjon commented:
@marionf thanks a lot for your feedback! :)

> 1. On the add/edit page, tooltips are missing for the following fields: name, displayed, description, category cover image, category thumbnail, menu thumbnails, meta title, meta description, meta keywords, friendly url and group acces.

These are replaced by help text below input.

> 2. French is my default language (in International > Localization)
I select French in the add/edit category page
I miss to fill a mandatory field (like name)
Click on Save
I have well the error message but the language selected is now English instead of French.

This is being fixed in https://github.com/PrestaShop/PrestaShop/pull/12214

> 3. When I edit a category name & save, it's not saved

Thanks, fixed. :+1: 

> 4. When I add a new category in the home category, I have 2 categories with position 1

Fixed. :+1: 

> 5. I am in multistore context and select the context "All shops"
When I click on add a new category, only the first shop is selected

Fixed. :+1: 

> 6. When I click to edit the "home" category I am redirected to the legacy page

Fixed. :+1: 

There are some weird behavior:
1. When you are adding new Root Category, "Description" field is not available, but when you edit existing Root Category (e.g. Home), description appears, what is the correct behavior?
2. When you click edit on "Root" category (as in 6th step in your feedback) it seems that the top most category should not have "Parent category" field, what do you think?

sarjon commented:
@matks i need some assistance for "Make rich text editor (TinyMCE) translatable" task, do you have any ideas? I can't make it toggleable for translatable inputs.

marionf commented:
Thank you for all the fixes :+1: 

> When you are adding new Root Category, "Description" field is not available, but when you edit existing Root Category (e.g. Home), description appears, what is the correct behavior?

It seems it was already the case in legacy page, but I have absolutely no idea why the description field isn't displayed

> When you click edit on "Root" category (as in 6th step in your feedback) it seems that the top most category should not have "Parent category" field, what do you think?

Yes, it was already like this in legacy page, the "parent category" tree displayed when you edit others categories isn't displayed for home category

![capture d ecran_1149](https://user-images.githubusercontent.com/13449658/54034109-59535e80-41b6-11e9-9c20-1b40f7ff2aa4.png)

sarjon committed:
fix dropdown hovering in action column

sarjon committed:
use groups provider instead of query to get default groups

sarjon committed:
remove default customers queyr from controller

sarjon committed:
remove Groups subdomain

sarjon committed:
fix missing arguments for category not found exception

sarjon committed:
fix naming for bulk action commands/handlers

sarjon committed:
removed CategoryStatus VO

sarjon committed:
improve choice tree rendering

sarjon committed:
cs fixer

sarjon committed:
added CleanHtml constraint to description

sarjon committed:
remove debug code

matks commented on classes/Category.php:
```diff
@@ -242,12 +242,11 @@ public function update($nullValues = false)
-                        $this->addPosition((int) Category::getLastPosition((int) $this->id_parent, (int) $idShop), (int) $idShop);
-                    }
+                foreach (Tools::getValue('checkBoxShopAsso_category') as $idAssoObject => $idShop) {
+                    $this->addPosition((int) Category::getLastPosition((int) $this->id_parent, (int) $idShop), (int) $idShop);
```
Is it a bug fix here ? Previous  `$row` variable was actually an `$idShop`, not an array ?

sarjon replied:
exactly. :+1:

marionf commented:
@sarjon 

1) When I edit a cateogry and put "displayed" parameter to No, I have the message "successful update" but it's not.
If I try to change it directly in the lisitng it's working.

2) Same thing for friendly URL field, it's not saved

3) If I select a category (like Accessories) and click on "Add new category", the parent cateogry "accessories" should be selected

4) When I edit "Home" or a root category, I have no description field. It was there in legacy, can you add it to keep the same behavior as in legacy ? As I don't know why there is no description field when you add a root category but it's here when you edit it.

5) In a multistore context, when I was clicking on the "root" link with legacy page, I was on the same page. I don't really know why, but I guess it's the wanted behavior

![capture d'écran_1160](https://user-images.githubusercontent.com/13449658/54546898-f24b5c00-49a4-11e9-836e-fd86a90a2e3b.png)

In symfony, I arrive on the root category and can edit it

![capture d'écran_1161](https://user-images.githubusercontent.com/13449658/54547030-422a2300-49a5-11e9-9883-939539fd0a10.png)

I am afraid people edit this "root" category and make mistakes

I tried to edit root category (changing name, meta title, meta description, parent category, group access, friendly url) and page is loading for more than 30 seconds, so I think it's safer to keep the same behavior as in legacy 

6) In fact, I should see this page only after adding a new root category (as in legacy)

![capture d'écran_1162](https://user-images.githubusercontent.com/13449658/54548440-368c2b80-49a8-11e9-98be-ba83ddaad9f1.png)

Currenly, after adding a new root category I am redirected to the Home category

![capture d'écran_1163](https://user-images.githubusercontent.com/13449658/54548485-515ea000-49a8-11e9-8930-176bcba90fa7.png)

7) After doing a lot of change of my shop, I have this issue when I try to edit a category

![capture d'écran_1164](https://user-images.githubusercontent.com/13449658/54552347-27a97700-49b0-11e9-82fb-572aae10f9ad.png)

Really sorry for this one, don't know how it happens and can't reproduce it, but if by chance, the error speaks to you :sweat_smile:

matks commented:
@marionf Thank you for the thorough bug report with lot of screnshots 😉

sarjon committed:
fix saving state for category

sarjon committed:
fix preselecting parent when adding new category

sarjon committed:
update parent category naming

sarjon committed:
fix missing description in root category add/edit actions

sarjon committed:
redirect to root category list after adding/editing root category

kpodemski commented:
hey @matks @sarjon 

i'm digging into 1.7.6 changes and i see that there are some breaking changes, right?

we no longer can to hook into dynamic hooks on migrated pages, like `actionAdminCategoriesFormModifier`, `actionAdminCategoriesControllerSaveAfter` right? i have dozen of modules working with those hooks, do you plan to document alternatives before 1.7.6 release?

matks commented:
Hi @kpodemski yes indeed, controller hooks are removed when we migrate a page because we rework how the controller works. However we made sure to provide new hooks for migrated pages, be it listing pages or add/edit pages.

A 1st documentation batch on this topic is already written and awaiting eternoendless approval at https://github.com/PrestaShop/docs/pull/184 and more will be added (you can see my list https://github.com/PrestaShop/docs/issues/198 and I think there's even more)

matks commented:
There's also https://github.com/PrestaShop/docs/pull/223 from mickaelandrieu for listing pages

kpodemski commented:
Great, i'll try to play with it a bit to see if all is clear :)

sarjon commented:
Hi @kpodemski,

you can take a look at `PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Handler\FormHandler` and hooks available in it. I think it's closed to `actionAdminCategoriesControllerSaveAfter`.

For form modification, take a look at `\PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder\FormBuilder` and hook available in it.

kpodemski commented:
Hey @eternoendless 

do you plan to release 1.7.6 beta/rc? just to give time for developers to handle breaking changes?

sarjon commented:
> do you plan to release 1.7.6 beta/rc? just to give time for developers to handle breaking changes?

I think you should use Gitter or dedicated issue, your question is not related to Categories PR. :)

sarjon commented:
@marionf huh :sweat_smile: thanks for feedback. :)

1. Fixed
2. Cannot reproduce, can you explain in which case it did not  save friendly URL?
3. Fixed
4. Fixed
5. Should user be able to edit `Root` category at all?
6. Fixed.
7. Hmmm, this error happens because category does not have name which is weird, because it's required field. Did you change your database directly? :thinking:

marionf commented:
Thnak you @sarjon for all fixes, there was a lot :)

> 2. Cannot reproduce, can you explain in which case it did not save friendly URL?

This is a small video, there is nothing special, I have no multistore, only one language

https://drive.google.com/open?id=1fCCnmyEpVWacvww6z4MuF7gZ4MKlsXJI

Same thing happens if I change the category name, the friendly url field is changed but not saved.

> 5. Should user be able to edit Root category at all?

We should keep the same behavior as in legacy: user can edit root category only after adding a new root category, he/she can click on edit. Otherwise, if I am on Home category and click on Root in the breadcrumb, I should stay on Home.

> 6. Fixed

I am still redirected to the Home category after adding a new root category

> 7. Hmmm, this error happens because category does not have name which is weird, because it's required field. Did you change your database directly? 

No, I didn't change it in database directly. I can't remember what I did, but yes category name was removed, just don't know how :sweat_smile: 


About the lack of tiny mce for description field, I just remembered that there is one in the maintenance page

![capture d'écran_1184](https://user-images.githubusercontent.com/13449658/54694874-4121fe80-4b29-11e9-8e7b-f508dd1181c3.png)

Don't know if you prefere integrate it in this PR or in another one ?

sarjon commented:
Thanks for reply @marionf, I'll double check.

> About the lack of tiny mce for description field, I just remembered that there is one in the maintenance page

It uses these weird tabs to switch language instead of drop-down next to input.

sarjon committed:
fix saving friendly url

sarjon committed:
fix link to root category in breadcrumbs

sarjon committed:
cs fixer

sarjon committed:
attept to fix breadcrumb for single root category

sarjon commented:
@marionf I managed to reproduce and fix saving friendly URL.

> We should keep the same behavior as in legacy: user can edit root category only after adding a new root category, he/she can click on edit. Otherwise, if I am on Home category and click on Root in the breadcrumb, I should stay on Home.

This behavior is so complex for such a simple use case. :/ I don't fully understand how and why it works they way it dies, but I've tried to replicate it in new page. However you are always able to access Root category if you change it in URL (same as in legacy). 

> I am still redirected to the Home category after adding a new root category

Weird, cannot reproduce it and I'm sure I've fixed it. :/

> No, I didn't change it in database directly.

Well, I hope we can fix it during freeze period. :sweat_smile:

marionf commented:
It's all good for me @sarjon 
Thank you very much, it was a huge one !

matks commented:
Thank you @sarjon for the PR and thank you @marionf for the great review !
